src/parser.rs

```diff
@@ -0,0 +1,287 @@
+        parse_binary_expression(-1)
+    }
+
+    fn parse_binary_expression(&self, cur_priority: u32) {
+        parse_expression();
+
+        // tries to read some binary operators
+        loop {
+            if let Op(op, op_priority, associativity) = self.try_peek_binary_op() {
+                if (op_priority > priority) break;
+                
+                self.read_op(op);
+                let expr2 = self.parse_binary_expression(op_priority);
+            }
+            else {
+                break;
+            }
+        }
+    }
+
+    fn parse_unary_expr(&self) {
+        // try read prefixOp
+        if self.try_peek_unary_op() {
+            self.read_op();
+
+            self.parse_unary_expr();
+        }
+        else {
+            self.parse_postfix_expression();
+        }
+    }
+
+    fn parse_postfix_expression(&self) {
+        self.parse_primary_expression();
+
+        loop {
+            if self.try_peek_op(Operator.Dot) {
+                self.read_identifier();
+
+                if (self.try_peek_op(Operator.LeftParen)) {
+                    self.parse_parenthesized_argument_list();
+                    // method call: EXPR.ident(arg1, arg2, ...)
+                }
+                else {
+                    // member reference: EXPR.ident
+                }
+            }
+            else if self.try_peek_op(Operator.LeftBrace) {
+                self.read_op(Operator.LeftBrace);
+                self.parse_expression();
+                self.read_op(Operator.RightBrace);
+                // array access: EXPR[EXPR]
+            }
+            else {
+                break;
+            }
+        }
+    }
+
+    fn parse_primary_expression(&self) {
+        if self.try_peek_keyword(Keyword.Null) {
+            self.read_keyword(Keyword.Null);
+        }
+        else if self.try_peek_keyword(Keyword.False) {
+            self.read_keyword(Keyword.False);
+        }
+        else if self.try_peek_keyword(Keyword.True) {
+            self.read_keyword(Keyword.True);
+        }
+        else if self.try_peek_integer() {
+            self.read_integer();
+        }
+        else if self.try_peek_identifier() {
+            self.read_identifier();
+
+            if self.try_read_op(Operator.LeftParen) {
+                // function call
+                self.parse_parenthesized_argument_list();
+            }
+            else {
+                // var ref
+            }
+        }
+        else if self.try_peek_keyword(Keyword.This) {
+            self.read_keyword(Keyword.This);
+        }
+        else if self.try_peek_op(Operator.LeftParen) {
+            // parenthesized expression
+            self.read_op(Operator.LeftParen);
+            self.parse_expression();
+            self.read(Operator.RightParen);
+        }
+        else if self.try_peek_keyword(Keyword.New) {
+            self.read_keyword(Keyword.New);
+
+            self.parse_basic_type();
+
+            if true {
```

Comment:
I'm not yet sure how to implement this nicely, but we cannot do this.

I would at least introduce another function `parse_new_expr`, since this is/will be non trivial logic.